### PR TITLE
fix(traverse): Image content w/ field_scale id

### PR DESCRIPTION
**Pull request by rpatterson**
Adds a test that reproduces a bug in the edge case where a user creates a new content
instance of the `Image` type that happens to have a `context.id` that matches the same
`{field}_{scale}` format that is introspected by the image scale traverser to try and
reproduce a bug in that logic.

Also adds a fix for the bug.  I went with a conditional at the top of
`plone.app.imaging.traverse:ImageTraverser.publishTraverse` because I'm not sure of all
the implications of just calling `self.fallback(request, name)` first, otherwise I would
have gone with that as it seems like the correct behavior to me.  So LMK if someone who
knows better than I also thinks that's the right approach and I'll make the change.

Due to [an unrelated
bug](https://github.com/plone/Products.CMFPlone/issues/2200#issuecomment-843632261), I
wasn't able to get this test to reproduce the bug on the latest applicable Plone that
still supports Archetypes, version 5.2 under Python 2.7.  I found it easiest to develop
the test case and fix against the branch used in the latest `buildout.coredev` branch
where `ATContentTypes` was the default, namely `4.3`.  Once we're done making changes to
the test and fix I'll also forward port this commit to the `5.0-5.2` branches.

----

This is probably my first contribution in years, so please LMK what's changed and what I
should do differently, thanks!

### CODE REVIEW

**Review by avoinea — APPROVED** [review]

**Comment by rpatterson** [comment]
Anyone else I should request review from, @mauritsvanrees @esteele ?

File: src/plone/app/imaging/profiles.zcml [hunk 1]
```diff
@@ -12,4 +12,12 @@
     provides="Products.GenericSetup.interfaces.EXTENSION"
     />
 
+  <genericsetup:registerProfile
```
Comment (mauritsvanrees) [inline src/plone/app/imaging/profiles.zcml:15]: This registers the testing profile always, instead of only during tests. You should move this to `testing.zcml`.

File: src/plone/app/imaging/tests/traversal.txt [hunk 2]
```diff
@@ -53,3 +53,47 @@ expression:
...
+  ... )
+  >>> shutil.copyfile(
+  ...     os.path.join(base.TESTS_PATH, "image.jpg"),
+  ...     test_image_scale_name,
+  ... )
+
+  >>> import __builtin__
+  >>> portal.portal_setup.runAllImportStepsFromProfile("plone.app.imaging:testing")
+  {...types: 'News Item Folder'...}
+  >>> with __builtin__.open(test_image_scale_name) as test_image_scale_name_opened:
```
Comment (mauritsvanrees) [inline src/plone/app/imaging/tests/traversal.txt:75]: Why do you need `__builtin__.open` here? Isn't `open` working?

**Review by mauritsvanrees — CHANGES_REQUESTED** [review]
Mostly okay, two inline comments that should be addressed.

What wasn't clear to me from the description is what the problem actually is and how to reproduce it. Apparently uploading *any* image fails on Plone 5.2 with Archetypes, but I cannot reproduce that: I can upload an image in the Plone Site root just fine.

Or is the uploading only a problem in this specific corner case? From reading the tests, the problem would be this:

- Create a Folder of a special type that has an image field. I think there is an add-on that uses archetypes.schemaextender to add a leadimage field, which could be used to the same effect, but I cannot find it in the GitHub collective.
- Maybe already upload an image in this field, but this may not be needed for reprocucing the bug.
- Add an Image (the content type) to this folder and upload an image called `image_mini.png`. This will work fine I think, because the id will be `image_mini.png`.
- Rename the image to `image_mini`.
- Now Plone will be confused between the image field and the image content. I don't know what happens then. Do you get an error when viewing the image content? Or when viewing the Folder? Can you not delete or rename the image content anymore?

You might argue that you should not be allowed to create content with the name `image_mini`, so an alternative could be to prevent this. I think there used to be a list of disallowed ids. Might be good to add `image_mini` to this. But with different  field names and scale names, the list could be long, and may need to be dynamic, and we are not going to do that at this point.

Anyway, the fix you do here in `traverse.py` seems good. Fine to merge if my two inline comments are addressed.

**Comment by mauritsvanrees** [comment]
O, when you create a PR, you should start the tests on Jenkins, which you should be able to do with your GitHub account. Usually you automatically get a comment about that from a bot called `mr.roboto`, and it ends up in the checks, but somehow that does not happen for this package. I have started a job now, and that one at least makes it to the checks:

https://jenkins.plone.org/job/pull-request-5.2-2.7/2031/

It will take a while to finish.
Luckily, since this is Archetypes, only one job is needed, for Python 2.

O, and it will fail. I tried it locally in the coredev buildout and got:

```
ImportError: No module named testbrowser
```

Ah, now I see: you made your PR against an old branch for Plone 4.3, like you said in your main comment.... That won't work when I test it against 5.2 of course.

With my Release Manager hat on: the only Plone version that is still supported, is 5.2. For older versions we will not make releases anymore. I am okay with occasionally making releases from old branches for individual packages, but they won't ever end up in an official release of Plone.

So if you really want, you could fix it on all branches that you want, and I could release them. But I am happy if you only fix it on master.

If you want to test locally with Plone coredev for 5.2: just run buildout, and then start the instance with `bin/instance-archetypes fg`. Then you can still create an Archetypes-based Plone Site. Choose the Advanced form for that, and select the profile for 'default Archetypes content' or something similar.

File: src/plone/app/imaging/profiles.zcml (hunk 1 above)
Comment (rpatterson) [reply src/plone/app/imaging/profiles.zcml:15]: Duh!  Sorry, will do.

File: src/plone/app/imaging/tests/traversal.txt (hunk 2 above)
Comment (rpatterson) [reply src/plone/app/imaging/tests/traversal.txt:75]: Because [something else is imported as `open`](https://github.com/plone/plone.app.imaging/blob/fix/traverse-subobject-scale-name/src/plone/app/imaging/tests/traversal.txt#L22).  I'm not really a fan of clobbering a built in like that but I went with this approach to keep my changes related to the traversal fix.

**Comment by rpatterson** [comment]
> Apparently uploading _any_ image fails on Plone 5.2 with Archetypes, but I cannot reproduce that: I can upload an image in the Plone Site root just fine.

Yeah, I link to that issue in the description.

> * Create a Folder of a special type that has an image field. I think there is an add-on that uses archetypes.schemaextender to add a leadimage field, which could be used to the same effect, but I cannot find it in the GitHub collective.

Yeah, it has to be in an instance of a folderish type with an `image` field.  In my experience, many, if not most, projects that define custom content types include a folderish type with an `image` field.  Indeed I was assigned to address this for Eau de Web because that project includes such a custom type.  So I think this is a case that needs to be addressed.

> * Maybe already upload an image in this field, but this may not be needed for reprocucing the bug.

IIRC, it doesn't  matter whether there's an actual image in the containing folder's `image` field.

> * Add an Image (the content type) to this folder and upload an image called `image_mini.png`. This will work fine I think, because the id will be `image_mini.png`.
> * Rename the image to `image_mini`.

We discovered the issue when uploading an image.  I could reproduce the issue by uploading an image with any name on the filesystem as long as I specified `image_mini` as the `id` on the upload form.

> * Now Plone will be confused between the image field and the image content. I don't know what happens then. Do you get an error when viewing the image content? Or when viewing the Folder? Can you not delete or rename the image content anymore?

I got an error when adding the `Image` instance but you could see the instance exists in the ZMI.  IIRC, when viewing the image the browser gets a text response containing the `repr(...)` of the traverser.

> You might argue that you should not be allowed to create content with the name `image_mini`, so an alternative could be to prevent this. I think there used to be a list of disallowed ids. Might be good to add `image_mini` to this. But with different field names and scale names, the list could be long, and may need to be dynamic, and we are not going to do that at this point.

I would argue otherwise.  I can easily imagine use cases where a user wants to and intends to upload a file called `image_thumb`  without an extension.  Also consider that users are allowed to define custom image scales, say `foo`, and developers can define custom field names, say `bar`, so it's entirely possible to end up with a case where trying to upload an image named `bar_foo` would reproduce the issue.  More than that having code that may *unintentionally* traverse to something else when a ZODB OFS object exists there just seems wrong.

**Comment by rpatterson** [comment]
> I have started a job now, and that one at least makes it to the checks:
> 
> https://jenkins.plone.org/job/pull-request-5.2-2.7/2031/

I don't see any way to see output at that link.  I also tried logging in via GitHub, but none of the links from that page lead to output that I can see.  Can you clarify?

> O, and it will fail. I tried it locally in the coredev buildout and got:
> 
> ```
> ImportError: No module named testbrowser
> ```

I don't see how any of my change could be related to that error.  AFAICT, I only re-used existing `Products.Five.testbrowser` usage in my tests.  Can you elaborate on what I should to to address this if anything?

> I am okay with occasionally making releases from old branches for individual packages, but they won't ever end up in an official release of Plone.

Ok, thanks much for the flexibility and your time!  I'll apply the change as appropriate to all branches once we've finished making revisions to this PR.  As far as releases, I think we might want another `1.0.x` release for the project this is affecting, but let me confirm that before you waste any of your time on it.  As far as other branches, once we've finished this PR and I'll apply this to all the other affected branches.  I'm agnostic regarding releases for other branches/versions so I'll leave that to you to decide.

**Comment by rpatterson** [comment]
@jenkins-plone-org please run jobs

File: src/plone/app/imaging/tests/traversal.txt (hunk 2 above)
Comment (mauritsvanrees) [reply src/plone/app/imaging/tests/traversal.txt:75]: That is ugly, but I understand it now.
After merge, we could change `from PIL.Image import open` to `from PIL.Image import open as open_image`.
But for now your approach keeps the PR diff smaller, which is good.

**Review by mauritsvanrees — APPROVED** [review]

**Comment by mauritsvanrees** [comment]
> I don't see any way to see output at that link.

If you hover over the 2031 menu item, it will become a dropdown where you can select Console output. This link:
https://jenkins.plone.org/job/pull-request-5.2-2.7/2031/console

But:

> I don't see how any of my change could be related to that error. AFAICT, I only re-used existing Products.Five.testbrowser usage in my tests. Can you elaborate on what I should to to address this if anything?

Nothing. Like I wrote after this in my answer: the Jenkins test that I started was for Plone 5.2, which is irrelevant for the branch you target. That branch is not tested anywhere anymore.

With that said, the change is obvious and I am going to assume that the tests pass, so I will merge. Thanks!

Do you plan on porting this to the master branch?

**Comment by rpatterson** [comment]
> Do you plan on porting this to the master branch?

Yes, now that we're done making revisions, expect a series of PRs to port this fix to all remaining relevant branches.

Also...

> As far as releases, I think we might want another 1.0.x release for the project this is affecting, but let me confirm that before you waste any of your time on it.

I confirmed we'd like a release of the `1.0.x` branch for Plone 4.3 so that we can update the version pin in our deployments being affected by this bug.  LMK here when it's done if you don't mind and thanks in advance!

**Comment by mauritsvanrees** [comment]
I have released 1.0.15.
